jax/image/scale.py

```diff
@@ -0,0 +1,194 @@
+
+from jax import lax
+from jax import numpy as jnp
+import numpy as np
+
+
+def _lanczos_kernel(radius: float):
+  def kernel(x: np.ndarray) -> np.ndarray:
+    x = np.abs(x)
+    y = radius * np.sin(np.pi * x) * np.sin(np.pi * x / radius)
+    with np.errstate(divide='ignore', invalid='ignore'):
+      out = y / (np.pi ** 2 * x ** 2)
+    out = np.where(x <= 1e-3, 1., out)
+    return np.where(x > radius, 0., out)
+
+  return radius, kernel
+
+
+def _triangle_kernel():
+  return 1., lambda x: np.maximum(0, 1 - np.abs(x))
+
+
+def _keys_cubic_kernel():
+  # http://ieeexplore.ieee.org/document/1163711/
+  # R. G. Keys. Cubic convolution interpolation for digital image processing.
+  # IEEE Transactions on Acoustics, Speech, and Signal Processing,
+  # 29(6):1153–1160, 1981.
+  def kernel(x: np.ndarray) -> np.ndarray:
+    x = np.abs(x)
+    out = ((1.5 * x - 2.5) * x) * x + 1.
+    out = np.where(x >= 1., ((-0.5* x + 2.5) * x - 4.) * x + 2., out)
+    return np.where(x >= 2., 0., out)
+  return 2., kernel
+
+
+def _compute_spans(input_size: int, output_size: int,
+                   scale: float, translate: float,
+                   kernel: Tuple[float, Callable[[np.ndarray], np.ndarray]],
+                   antialias: bool) -> Tuple[np.ndarray, np.ndarray]:
+  """
+  Computes the locations and weights of the spans of a 1D input image.
+  Returns:
+    A `(starts, weights)` tuple of ndarrays, where `starts` has shape
+    `[output_size]` and `weights` has shape `[output_size, span_size]`.
+  """
+  radius, kernel_fn = kernel
+  inv_scale = 1. / scale
+  kernel_scale = max(inv_scale, 1.) if antialias else 1.
+  span_size = min(2 * int(math.ceil(radius * kernel_scale)) + 1, input_size)
+
+  sample_f = (np.arange(output_size) + 0.5) * inv_scale * (1. - translate)
+  span_start = np.ceil(sample_f - radius * kernel_scale - 0.5)
+  span_start = np.clip(span_start, 0, input_size - span_size)
+  kernel_pos = (span_start - sample_f)[:, None] + np.arange(span_size) + 0.5
+  weight = kernel_fn(kernel_pos / kernel_scale)
+  total_weight_sum = np.sum(weight, axis=1, keepdims=True)
+  weights = np.where(
+      np.abs(total_weight_sum) > 1000. * np.finfo(np.float32).min,
+      weight / total_weight_sum, 0)
+  return span_start.astype(np.int32), weights
+
+
+def _scale_and_translate(x, output_shape, scale, translate, kernel,
+                         antialias):
+  input_shape = x.shape
+  assert len(input_shape) == len(output_shape)
+  assert len(input_shape) == len(scale)
+  assert len(input_shape) == len(translate)
+  spatial_dims = np.nonzero(
+      np.not_equal(input_shape, output_shape) |
+      np.not_equal(scale, 1) |
+      np.not_equal(translate, 0))[0]
+  if len(spatial_dims) == 0:
+    return x
+  output_spatial_shape = tuple(np.array(output_shape)[spatial_dims])
+  indices = []
+  dim_weights = []
+  slice_shape = list(input_shape)
+  in_names = string.ascii_lowercase[:len(output_shape) + len(spatial_dims)]
```

Comment:
`einsum` also supports an alternative syntax that doesn't require string parsing. That might be slightly cleaner here?
```
a = np.arange(60.).reshape(3,4,5)
b = np.arange(24.).reshape(4,3,2)
np.einsum('ijk,jil->kl', a, b)
array([[4400., 4730.],
       [4532., 4874.],
       [4664., 5018.],
       [4796., 5162.],
       [4928., 5306.]])
np.einsum(a, [0,1,2], b, [1,0,3], [2,3])  # equivalent to the above
array([[4400., 4730.],
       [4532., 4874.],
       [4664., 5018.],
       [4796., 5162.],
       [4928., 5306.]])
```

(not really important given that this is already working!)

Reply:
Ah, I was looking for that. It's completely undocumented, I think? https://numpy.org/doc/stable/reference/generated/numpy.einsum.html

That said I felt dirty manipulating strings, so I will try this.

Reply:
It's definitely under-documented, but it's nicer for programmatic use

(and it works in JAX, too, because we copied the argument parsing logic from NumPy)

Reply:
For posterity, we implemented both forms of argument parsing "by hand" in [Autograd's einsum vjp](https://github.com/HIPS/autograd/blob/c6f630a5ec18bd30f1485bc0dbbccb8664c77510/autograd/numpy/numpy_vjps.py#L591-L626), and the non-string form was super easy to handle. I suspect it's meant for easy programmatic manipulation, whereas the strings can be easier for humans. (I do love how undocumented it is though, except for one sentence in the docstring!)